clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/provider/agent/AmazonInstanceTypeCachingAgent.java

```diff
@@ -0,0 +1,196 @@
+
+        //we have enough from a previous request to not re-request if the etag is unchanged..
+        HttpResponse headResponse = httpClient.execute(pricingHost, new HttpHead(pricingUri));
+        EntityUtils.consumeQuietly(headResponse.getEntity());
+        if (headResponse.getStatusLine().getStatusCode() != 200) {
+          throw new Exception("failed to read instance type metadata for " + region + ": " + headResponse.getStatusLine().toString());
+        }
+
+        Optional<String> etag = getEtagHeader(headResponse);
+
+        if (etag.filter(metadataAttributes.etag::equals).isPresent()) {
+          instanceTypes = metadataAttributes.cachedInstanceTypes;
+        }
+      }
+      if (instanceTypes == null) {
+        HttpResponse getResponse = httpClient.execute(pricingHost, new HttpGet(pricingUri));
+        if (getResponse.getStatusLine().getStatusCode() != 200) {
+          EntityUtils.consumeQuietly(getResponse.getEntity());
+          throw new Exception("failed to read instance type data for " + region + ": " + getResponse.getStatusLine().toString());
+        }
+        Optional<String> etag = getEtagHeader(getResponse);
+        HttpEntity entity = getResponse.getEntity();
+        instanceTypes = fromStream(entity.getContent());
+        EntityUtils.consumeQuietly(entity);
+        if (etag.isPresent()) {
+          metadataAttributes = new MetadataAttributes();
+          metadataAttributes.etag = etag.get();
+          metadataAttributes.cachedInstanceTypes = new HashSet<>(instanceTypes);
+          metadata = new DefaultCacheData("metadata", objectMapper.convertValue(metadataAttributes, ATTRIBUTES), Collections.emptyMap());
+        } else {
+          metadata = null;
+        }
+      }
+      Map<String, Collection<String>> evictions = new HashMap<>();
+      Map<String, Collection<CacheData>> cacheResults = new HashMap<>();
+      List<CacheData> instanceTypeData = new LinkedList<>();
+      cacheResults.put(Keys.Namespace.INSTANCE_TYPES.getNs(), instanceTypeData);
+      if (metadata != null) {
+        cacheResults.put(getAgentType(), Collections.singleton(metadata));
+      } else {
+        evictions.put(getAgentType(), Collections.singleton("metadata"));
+      }
+
+      for (String instanceType : instanceTypes) {
+        for (String account : matchingAccounts) {
+          Map<String, Object> instanceTypeAttributes = new HashMap<>();
+          instanceTypeAttributes.put("account", account);
+          instanceTypeAttributes.put("region", region);
+          instanceTypeAttributes.put("name", instanceType);
+          instanceTypeData.add(new DefaultCacheData(Keys.getInstanceTypeKey(instanceType, region, account), instanceTypeAttributes, Collections.emptyMap()));
+        }
+      }
+
+      return new DefaultCacheResult(cacheResults, evictions);
+    } catch (Exception ex) {
+      throw new RuntimeException(ex);
```

Comment:
do we really want to make every exception sneaky here?

Reply:
it's effectively what most other caching agents to, just leveraging loosy groovy to do so.

The scheduler will log and reschedule the agent if it fails 

If there was a specifically handleable exception in here I'd be all for handling it (if find we need retries on the http call or anything)